### PR TITLE
feature(rpc): Add basic long polling support to the `getblocktemplate` RPC

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR implements basic long polling support, checking for changes by re-fetching from the sync status, state and mempool.

Part of ticket #5720.

### Complex Code or Requirements

We need to hold the RPC request open until the response will change. This means running a loop and a `select!{}` in a future.

## Solution

Implement inefficient long polling by checking the state/mempool/syncer, waiting 5 seconds, then checking it again. Also wait for the max time.

Related changes:
- Change `impl From<LongPollInput> for LongPollId` into a named method
- Rename the `long_poll_id` field
- Cleanup some comments

## Review

This is ready for review, I'm doing some manual testing now. Automated testing will happen in the next PR.

Efficiency and refactors are out of scope for this PR, I'd like to focus on getting some kind of working implementation merged.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- medium priority: check for state changes efficiently
- low priority: check for sync status and mempool changes efficiently

### CODE REVIEW

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (arya2) [inline zebra-rpc/src/methods/get_block_template_rpcs.rs:374]: We should use `mempool_transaction_receiver` to check if something is inserted into the mempool. https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/commands/start.rs#L163

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (arya2) [inline zebra-rpc/src/methods/get_block_template_rpcs.rs:340]: I think we should block this method here by awaiting `wait_for_max_time`, a changed `latest_chain_tip`, or a changed `mempool_transaction_receiver`, and leave the rest of this function as it is.

Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:340]: I'm not sure if this would work the way it is meant to?

The long polling spec says that we need to return immediately if the client long poll ID is different from the current server long poll ID:
> This request SHOULD NOT be processed nor answered by the server until it wishes to replace the current block data as identified by the "longpollid".

https://en.bitcoin.it/wiki/BIP_0022#Optional:_Long_Polling

Using watch channels is also out of scope for this PR, see the TODOs below.

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:374]: Using watch channels is out of scope for this PR, see the TODOs.

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (arya2) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:340]: > I'm not sure if this would work the way it is meant to?

Ah, you're right, it needs to check that the long poll id is the same first.

**Comment by teor2345** [comment]
This is ready for review, I'm doing some manual testing now. Automated testing will happen in the next PR.

**Comment by teor2345** [comment]
A bunch of google cloud jobs seemed to have failed setup with:
> ssh: connect to host 104.198.199.90 port 22: Connection timed out

https://github.com/ZcashFoundation/zebra/actions/runs/3689513976/jobs/6246166936#step:11:78

**Comment by teor2345** [comment]
@mergifyio update

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (arya2) [inline zebra-rpc/src/methods/get_block_template_rpcs.rs:383]: Do we want to use the max time from the client's long poll id?

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (arya2) [inline zebra-rpc/src/methods/get_block_template_rpcs.rs:392]: I think this would return a template with a `max_time` in the past if `max_time_reached` is true but the long poll id is the same.

**Review by arya2 — COMMENTED** [review]
This looks like it should work.

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:383]: I think that would make the next poll ignore any changes in the block header due to a new `max_time` from the state?

The `max_time` can change if the local clock is close to or past the minimum difficulty time on testnet. (It doesn't matter on mainnet.)

Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:383]: All the gory details are here:
https://github.com/ZcashFoundation/zebra/blob/4fb2417adc70f57f42e9df4c8467279058b743d7/zebra-state/src/service/read/difficulty.rs#L263-L276

File: zebra-rpc/src/methods/get_block_template_rpcs.rs
Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:392]: Consensus rules can't depend on the local node time, because they need to be global. So it's definitely unusual, but I think it's still valid?

`max_time` is clamped to 90 minutes later than the `median-time-past` of recent blocks, based on this consensus rule:
https://github.com/ZcashFoundation/zebra/blob/4fb2417adc70f57f42e9df4c8467279058b743d7/zebra-state/src/service/read/difficulty.rs#L213-L219

(There is a standard 2 hour rule for the local node time, but standard rules can vary between implementations or configurations, so we don't use it to construct the block template.)

Comment (arya2) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs:392]: Ah, okay, so after the max time is reached there can't be any more valid changes to the template but the block can still be mined. This looks right then.

**Review by arya2 — APPROVED** [review]

**Comment by teor2345** [comment]
@arya2 just checking why you marked this PR as "do-not-merge"?

It would be easier for me if this was part of the `main` branch and I didn't have to rebase my changes for PR #5862.

I'm also happy to sort out tests and any fixes in the next PR.

**Comment by oxarbitrage** [comment]
I think is in case someone else wanted to review as @dconnolly is in the list as well.

**Comment by oxarbitrage** [comment]
> I think is in case someone else wanted to review as @dconnolly is in the list as well.

But i think one approval is ok here so we can just remove the do not merge tag.

**Comment by arya2** [comment]
> @arya2 just checking why you marked this PR as "do-not-merge"?

Just in case we wanted a second reviewer. I'll remove the label.

**Comment by teor2345** [comment]
> > @arya2 just checking why you marked this PR as "do-not-merge"?
> 
> Just in case we wanted a second reviewer. I'll remove the label.

Always happy to get a second review, but I try not to block unless someone has committed to doing a review soon.

**Pull request merged**
